checker/src/body_visitor.rs

```diff
@@ -1386,88 +1386,44 @@ impl<'analysis, 'compilation, 'tcx, E> BodyVisitor<'analysis, 'compilation, 'tcx
-                        // fall through, the target path is unique
-                    } else {
-                        self.weak_updates(qualifier, &source_path, |v| value.equals(v));
-                        // and now fall through for a strong update of target_path
-                    }
-                }
-                PathSelector::Slice(count) => {
-                    // if the count is known at this point, expand it like a pattern.
-                    if let Expression::CompileTimeConstant(ConstantDomain::U128(val)) =
-                        &count.expression
-                    {
-                        self.expand_slice(
-                            &qualifier,
-                            &source_path,
-                            0,
-                            *val as u32,
-                            |_self, target_path, source_path| {
-                                _self.copy_or_move_elements(
-                                    target_path,
-                                    source_path,
-                                    root_rustc_type,
-                                    move_elements,
-                                )
-                            },
-                        );
-                    } else {
-                        // Any that that might get assigned to by this slice need to get weakened.
-                        // That is, the value has to become a condition based on index < count.
-                        self.weaken_paths_that_index(qualifier, count);
-                    }
-                    // fall through
-                }
-                _ => {
-                    // fall through
-                }
-            }
-        }
-
-        // Assigning to a fixed length array is like a pattern
-        if let TyKind::Array(ty, len) = &root_rustc_type.kind {
-            let param_env = self.type_visitor.get_param_env();
-            let len = len
-                .try_eval_usize(self.tcx, param_env)
-                .expect("Array to have constant len");
-
-            self.expand_slice(
-                &target_path,
-                &source_path,
-                0,
-                len as u32,
-                |_self, target_path, source_path| {
-                    _self.copy_or_move_elements(target_path, source_path, ty, move_elements)
-                },
-            );
+        // If the pattern is a known size, copy_of_move_elements is called recursively
+        // on each expansion of the target pattern and try_expand_target_pattern returns true.
+        // If not, all of the paths that might match the pattern are weakly updated to account
+        // for the possibility that this assignment might update them (or not).
+        let expanded_target_pattern = self.try_expand_target_pattern(
+            &target_path,
+            &source_path,
+            root_rustc_type,
+            |_self, target_path, source_path, root_rustc_type| {
```

Comment:
The name `root_rustc_type` here might be a bit misleading. Maybe `elem_rustc_type`?

Reply:
It is a bit subtle, but from the context of the call to _self.copy_or_move_elements, the parameter named root_rutstc_type is the type of whatever kind of value is rooted at source_path, so elem_rustc_type will be misleading as well. Also, when target pattern is expanded to include conditional address expressions, it will not even be true that recursive calls will be to elements of an array.